### PR TITLE
ODataReadOnlySupport

### PR DESCRIPTION
**Pull request by cristinamanum**
Add ODataReadOnlySupport

### CODE REVIEW

File: src/NuGetGallery.Core/Entities/ReadOnlyEntitiesContext.cs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+using System.Linq;
+using NuGet.Services.Entities;
+
+namespace NuGetGallery
+{
+    public class ReadOnlyEntitiesContext : IReadOnlyEntitiesContext
+    {
+        private readonly EntitiesContext _entitiesContext; 
+        
+        public ReadOnlyEntitiesContext(DbConnection connection)
```
Comment (skofman1) [inline src/NuGetGallery.Core/Entities/ReadOnlyEntitiesContext.cs]: Is there value in making this testable? If so consider using GalleryDbContextFactory

File: src/NuGetGallery.Core/Entities/EntitiesContext.cs [hunk 2]
```diff
@@ -53,7 +53,7 @@ public EntitiesContext(string connectionString, bool readOnly)
         public EntitiesContext(DbConnection connection, bool readOnly)
             : base(connection, contextOwnsConnection: true)
         {
-            ReadOnly = readOnly;
+           ReadOnly = readOnly;
```
Comment (joelverhagen) [inline src/NuGetGallery.Core/Entities/EntitiesContext.cs]: 4 spaces for indentation

File: src/NuGetGallery/Controllers/ODataV2FeedController.cs [hunk 3]
```diff
@@ -597,5 +603,14 @@ where isSemVerLevelCompliant(p)
 
             return updates;
         }
+
+        private IQueryable<Package> GetAll()
```
Comment (skofman1) [inline src/NuGetGallery/Controllers/ODataV2FeedController.cs]: Is there a test to verify this code?

File: src/NuGetGallery/App_Start/DefaultDependenciesModule.cs [hunk 4]
```diff
@@ -562,6 +564,28 @@ private static DbConnection CreateDbConnection(ISqlConnectionFactory connectionF
...
+            ISecretInjector secretInjector)
+        {
+            var galleryDbReadOnlyReplicaConnectionFactory = CreateDbConnectionFactory(
+                diagnostics,
+                nameof(ReadOnlyEntitiesContext),
+                configuration.Current.SqlReadOnlyReplicaConnectionString ?? configuration.Current.SqlConnectionString,
+                secretInjector);
+
+            builder.Register(c => new ReadOnlyEntitiesContext(CreateDbConnection(galleryDbReadOnlyReplicaConnectionFactory)))
+                .AsSelf()
```
Comment (joelverhagen) [inline src/NuGetGallery/App_Start/DefaultDependenciesModule.cs]: An we not register `ReadOnlyEntitiesContext`? IMO you should only inject interfaces.

File: src/NuGetGallery/App_Start/DefaultDependenciesModule.cs [hunk 5]
```diff
@@ -562,6 +564,28 @@ private static DbConnection CreateDbConnection(ISqlConnectionFactory connectionF
...
+                .As<IReadOnlyEntitiesContext>()
+                .InstancePerLifetimeScope();
+
+            builder.RegisterType<ReadOnlyEntityRepository<Package>>()
+                .AsSelf()
```
Comment (joelverhagen) [inline src/NuGetGallery/App_Start/DefaultDependenciesModule.cs]: Same, I don't see why a service would import `ReadOnlyEntityRepository<Package>`

File: src/NuGetGallery.Core/Entities/IReadOnlyEntitiesContext.cs [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+using System.Data.Entity;
+using NuGet.Services.Entities;
+
+namespace NuGetGallery
+{
+    public interface IReadOnlyEntitiesContext : IDisposable
+    {
+        DbSet<Package> Packages { get; set; }
+
+        DbSet<T> Set<T>() where T : class;
```
Comment (agr) [inline src/NuGetGallery.Core/Entities/IReadOnlyEntitiesContext.cs]: Need to document the purpose of the method. Also name seems confusing: it's called `Set` but does not set anything, instead it returns something.

File: src/NuGetGallery/Controllers/ODataV1FeedController.cs [hunk 7]
```diff
@@ -281,5 +287,14 @@ public IHttpActionResult GetPropertyFromPackages(string propertyName, string id,
             var searchResults = await Search(options, searchTerm, targetFramework);
             return searchResults.FormattedAsCountResult<V1FeedPackage>();
         }
+
+        private IQueryable<Package> GetAll()
+        {
+            if (_featureFlagService.IsODataDatabaseReadOnlyEnabled())
```
Comment (agr) [inline src/NuGetGallery/Controllers/ODataV1FeedController.cs]: This should be resolved by DI, isn't it?

File: src/NuGetGallery/App_Start/DefaultDependenciesModule.cs [hunk 8]
```diff
@@ -562,6 +564,28 @@ private static DbConnection CreateDbConnection(ISqlConnectionFactory connectionF
...
+            ConfigurationService configuration,
+            ISecretInjector secretInjector)
+        {
+            var galleryDbReadOnlyReplicaConnectionFactory = CreateDbConnectionFactory(
+                diagnostics,
+                nameof(ReadOnlyEntitiesContext),
+                configuration.Current.SqlReadOnlyReplicaConnectionString ?? configuration.Current.SqlConnectionString,
+                secretInjector);
+
+            builder.Register(c => new ReadOnlyEntitiesContext(CreateDbConnection(galleryDbReadOnlyReplicaConnectionFactory)))
```
Comment (agr) [inline src/NuGetGallery/App_Start/DefaultDependenciesModule.cs]: DI should fallback to providing `EntitiesContext` instances when `IReadOnlyEntitiesContext` is requested when feature flag is disabled, so we would make this choice in a single location. Same for `ReadOnlyEntityRepository<Package>` below.

File: src/NuGetGallery.Core/Entities/ReadOnlyEntitiesContext.cs (hunk 1 above)
Comment (cristinamanum) [reply src/NuGetGallery.Core/Entities/ReadOnlyEntitiesContext.cs]: I do not think that it is worth. It is only redirecting to the EntitiesContext.

File: src/NuGetGallery.Core/Entities/EntitiesContext.cs (hunk 2 above)
Comment (cristinamanum) [reply src/NuGetGallery.Core/Entities/EntitiesContext.cs]: Fixed.

File: src/NuGetGallery/Controllers/ODataV1FeedController.cs (hunk 7 above)
Comment (cristinamanum) [reply src/NuGetGallery/Controllers/ODataV1FeedController.cs]: After the feature flag will be removed this method will go away. If the feature flag is not on the read-write will be used.

File: src/NuGetGallery/App_Start/DefaultDependenciesModule.cs (hunk 8 above)
Comment (cristinamanum) [reply src/NuGetGallery/App_Start/DefaultDependenciesModule.cs]: The feature flags may be changed multiple times after startup. For this reason the decision is done in the controller.

**Pull request merged**